drivers/hmis/app/models/hmis/hud/base.rb

```diff
@@ -50,4 +52,19 @@ def skip_validations
   def required_fields
     @required_fields ||= []
   end
+
+  private def ensure_id
+    return if send(self.class.hud_key).present? # Don't overwrite the ID if we already have one
+
+    assign_attributes(self.class.hud_key => self.class.generate_uuid)
+  end
+
+  # Let Rails update the HUD timestamps
+  def self.timestamp_attributes_for_create
+    super << 'DateCreated'
+  end
+
+  def self.timestamp_attributes_for_update
+    super << 'DateUpdated'
+  end
```

Comment:
@awisema  do you know if there's a way to call these on validation? Otherwise we get validation errors about them missing when we first call `valid?` on a new unsaved object.

I changed `before_save :ensure_id` to `before_validation :ensure_id` to fix a similar issue for IDs.

Reply:
Short answer -- for Rails, those dates represent when the database record is created/updated. so a unsaved object won't have them.

I see that you have a method `skip_validations` in `Hmis::Hud::Base`, do these need to exist for other reasons, or can you just skip them? If not,, we can add `after_initialize` hooks to the base class to set initial values for them -- I believe that if you set the create timestamp rails honors it, but the update timestamp will get changed...

Reply:
can you just add a `before_validation :ensure_timestamps` and have something like
```
private def ensure_timestamps
  self.date_created ||= DateTime.current
  self.date_updated ||= DateTime.current
end
```

Reply:
I think Sandy's approach makes sense to skip them for validation and have Rails deal with setting them. In all the mutations right now we are manually setting date_created and date_updated. I think this approach of using `timestamp_attributes_for_update` and `timestamp_attributes_for_create` means we can get rid of that. If we are 100% sure that Rails will set them, then we shouldn't need to validate their presence before saving/updating..

Reply:
https://github.com/greenriver/hmis-warehouse/pull/2788/commits/6d2567cdecd36001b389ec84f64335474deacab1 this is a relief. thank you @awisema 🙏